File: utils/hyperloglog/hll-err.rb
# hll-err.rb - Copyright (C) 2014-Present Redis Ltd.
#
# Licensed under your choice of (a) the Redis Source Available License 2.0
# (RSALv2); or (b) the Server Side Public License v1 (SSPLv1); or (c) the
# GNU Affero General Public License v3 (AGPLv3).
#
# Check error of HyperLogLog Redis implementation for different set sizes.

require 'rubygems'
require 'redis'
require 'digest/sha1'

r = Redis.new
r.del('hll')
i = 0
while true do
    100.times {
        elements = []
        1000.times {
            ele = Digest::SHA1.hexdigest(i.to_s)
            elements << ele
            i += 1
        }
        r.pfadd('hll',elements)
    }
    approx = r.pfcount('hll')
    abs_err = (approx-i).abs
    rel_err = 100.to_f*abs_err/i
    puts "#{i} vs #{approx}: #{rel_err}%"
end

File: utils/hyperloglog/hll-err.py
import hashlib
import random
import redis
import sys


def main(randomize=False):
    KEY = 'hllpy'
    r = redis.client.Redis()
    r.delete(KEY)
    jcnt = 0
    num_chunks = 100
    chunk_size = 1000
    while True:
        for j in xrange(num_chunks):
            if randomize:
                elements = [hashlib.sha1(str(i + jcnt) + '.' + str(random.randint(0, chunk_size))).hexdigest()
                            for i in xrange(chunk_size)]
            else:
                elements = [hashlib.sha1(str(i + jcnt)).hexdigest() for i in xrange(chunk_size)]
            r.execute_command('PFADD', KEY, *elements)
            jcnt += chunk_size
        approx = r.execute_command('PFCOUNT', KEY)
        err_pc = 100.0 * abs(jcnt - approx) / jcnt
        print '{} vs {}: {:5.4f}%'.format(jcnt, approx, err_pc)


if __name__ == '__main__':
    randomize = True if sys.argv[1:] else False
    try:
        main(randomize)
    except KeyboardInterrupt:
        pass
